File: yogi-cpp/test/yogi/timer_test.cc
/*
 * This file is part of the Yogi Framework
 * https://github.com/yohummus/yogi-framework.
 *
 * Copyright (c) 2020 Johannes Bergmann.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <test/common.h>

#include <yogi_core.h>

class TimerTest : public Test {
 protected:
  TimerPtr create_timer() {
    MOCK_TimerCreate([](void** timer, void*) {
      *timer = kPointer;
      return YOGI_OK;
    });

    return Timer::create(create_context());
  }
};

TEST_F(TimerTest, Create) {
  MOCK_TimerCreate([](void** timer, void* context) {
    EXPECT_NE(timer, nullptr);
    EXPECT_EQ(context, kPointer);
    return YOGI_OK;
  });
  Timer::create(create_context());
}

TEST_F(TimerTest, CreateError) {
  MOCK_TimerCreate([](void**, void*) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(Timer::create(create_context()), FailureException);
}

TEST_F(TimerTest, StartAsync) {
  auto timer = create_timer();

  bool called = false;
  auto fn     = [&](const Result& res) {
    EXPECT_EQ(res.error_code(), ErrorCode::kWrongObjectType);
    called = true;
  };

  MOCK_TimerStartAsync([](void* timer, long long duration, void (*fn)(int res, void* userarg), void* userarg) {
    EXPECT_EQ(timer, kPointer);
    EXPECT_EQ(duration, 1234);
    EXPECT_NE(fn, nullptr);
    EXPECT_NE(userarg, nullptr);
    fn(YOGI_ERR_WRONG_OBJECT_TYPE, userarg);
    return YOGI_OK;
  });
  timer->start_async(Duration::from_nanoseconds(1234), fn);
  EXPECT_TRUE(called);
}

TEST_F(TimerTest, StartAsyncError) {
  auto timer = create_timer();

  MOCK_TimerStartAsync([](void*, long long, void (*)(int, void*), void*) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(timer->start_async(Duration::from_nanoseconds(1234), {}), FailureException);
}

TEST_F(TimerTest, Cancel) {
  auto timer = create_timer();

  MOCK_TimerCancel([](void* timer) {
    EXPECT_EQ(timer, kPointer);
    return YOGI_OK;
  });
  EXPECT_TRUE(timer->cancel());

  MOCK_TimerCancel([](void*) { return YOGI_ERR_TIMER_EXPIRED; });
  EXPECT_FALSE(timer->cancel());
}

TEST_F(TimerTest, CancelError) {
  auto timer = create_timer();

  MOCK_TimerCancel([](void*) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(timer->cancel(), FailureException);
}

File: yogi-dotnet/test/TimerTest.cs
/*
 * This file is part of the Yogi Framework
 * https://github.com/yohummus/yogi-framework.
 *
 * Copyright (c) 2020 Johannes Bergmann.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

using System;
using Xunit;

namespace test
{
    public class TimerTest : TestCase
    {
        Yogi.Timer timer = CreateTimer();

        public static Yogi.Timer CreateTimer()
        {
            MOCK_TimerCreate((ref IntPtr timer, IntPtr context) =>
            {
                timer = pointer;
                return (int)Yogi.ErrorCode.Ok;
            });

            return new Yogi.Timer(CreateContext());
        }

        [Fact]
        public void Create()
        {
            MOCK_TimerCreate((ref IntPtr timer, IntPtr context) =>
            {
                Assert.Equal(context, pointer);
                timer = pointer;
                return (int)Yogi.ErrorCode.Ok;
            });

            new Yogi.Timer(CreateContext());
        }

        [Fact]
        public void CreateError()
        {
            MOCK_TimerCreate((ref IntPtr timer, IntPtr context) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.Throws<Yogi.FailureException>(() =>
            {
                new Yogi.Timer(CreateContext());
            });
        }

        [Fact]
        public void StartAsync()
        {
            bool called = false;

            Yogi.Timer.HandlerFnDelegate fn = (Yogi.Result result) =>
            {
                Assert.IsType<Yogi.Success>(result);
                Assert.Equal(Yogi.ErrorCode.Ok, result.ErrorCode);
                called = true;
            };

            MOCK_TimerStartAsync((IntPtr timer, long duration, TimerStartAsyncFnDelegate fn_, IntPtr userarg) =>
            {
                Assert.Equal(pointer, timer);
                Assert.Equal(1234, duration);
                Assert.NotNull(fn_);
                fn_((int)Yogi.ErrorCode.Ok, userarg);
                return (int)Yogi.ErrorCode.Ok;
            });

            timer.StartAsync(Yogi.Duration.FromNanoseconds(1234), fn);
            Assert.True(called);
        }

        [Fact]
        public void StartAsyncCallbackCancel()
        {
            bool called = false;

            Yogi.Timer.HandlerFnDelegate fn = (Yogi.Result result) =>
            {
                Assert.IsType<Yogi.Failure>(result);
                Assert.Equal(Yogi.ErrorCode.Canceled, result.ErrorCode);
                called = true;
            };

            MOCK_TimerStartAsync((IntPtr timer, long duration, TimerStartAsyncFnDelegate fn_, IntPtr userarg) =>
            {
                Assert.NotNull(fn_);
                fn_((int)Yogi.ErrorCode.Canceled, userarg);
                return (int)Yogi.ErrorCode.Ok;
            });

            timer.StartAsync(Yogi.Duration.FromNanoseconds(1234), fn);
            Assert.True(called);
        }

        [Fact]
        public void StartSyncError()
        {
            Yogi.Timer.HandlerFnDelegate fn = (Yogi.Result result) =>
            {
            };

            MOCK_TimerStartAsync((IntPtr timer, long duration, TimerStartAsyncFnDelegate fn_, IntPtr userarg) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.Throws<Yogi.FailureException>(() =>
            {
                timer.StartAsync(Yogi.Duration.FromNanoseconds(1234), fn);
            });
        }

        [Fact]
        public void Cancel()
        {
            MOCK_TimerCancel((IntPtr timer) =>
            {
                Assert.Equal(timer, pointer);
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.True(timer.Cancel());

            MOCK_TimerCancel((IntPtr timer) =>
            {
                Assert.Equal(timer, pointer);
                return (int)Yogi.ErrorCode.TimerExpired;
            });

            Assert.False(timer.Cancel());
        }

        [Fact]
        public void CancelError()
        {
            MOCK_TimerCancel((IntPtr timer) =>
            {
                Assert.Equal(timer, pointer);
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.Throws<Yogi.FailureException>(() =>
            {
                timer.Cancel();
            });
        }
    }
}
